File: src/transfers/YouthAcademy.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;
using ElevenLegends.Simulation;

namespace ElevenLegends.Transfers;

/// <summary>
/// Generates youth prospects for the academy recruitment system (card-based gacha).
/// </summary>
public static class YouthAcademy
{
    private static readonly Position[] YouthPositions =
    [
        Position.GK, Position.CB, Position.LB, Position.RB,
        Position.CM, Position.CDM, Position.CAM,
        Position.LW, Position.RW, Position.ST, Position.CF
    ];

    /// <summary>
    /// Generates 3 youth prospect cards. Player picks 1.
    /// </summary>
    public static List<(Player Prospect, decimal Fee)> GenerateProspects(
        IRng rng, string country, int nextPlayerId)
    {
        var names = GetNamePool(country);
        var prospects = new List<(Player, decimal)>();

        for (int i = 0; i < 3; i++)
        {
            int age = rng.NextInt(16, 19);
            var pos = YouthPositions[rng.NextInt(0, YouthPositions.Length - 1)];

            int baseAttr = rng.NextInt(30, 55);
            int variance = 12;

            string firstName = names.FirstNames[rng.NextInt(0, names.FirstNames.Count - 1)];
            string lastName = names.LastNames[rng.NextInt(0, names.LastNames.Count - 1)];
            string name = $"{firstName[0]}. {lastName}";

            var attrs = GenerateYouthAttributes(rng, pos, baseAttr, variance);

            var prospect = new Player
            {
                Id = nextPlayerId + i,
                Name = name,
                PrimaryPosition = pos,
                Age = age,
                Morale = rng.NextInt(50, 70),
                Chemistry = rng.NextInt(30, 50),
                Attributes = attrs
            };

            float ovr = pos == Position.GK ? attrs.GoalkeeperOverall : attrs.OutfieldOverall;
            decimal fee = Math.Round((decimal)(ovr * 200) / 1000m) * 1000m;
            fee = Math.Max(5_000m, Math.Min(15_000m, fee));

            prospects.Add((prospect, fee));
        }

        return prospects;
    }

    /// <summary>
    /// Returns the maximum player ID across all clubs (for generating new IDs).
    /// </summary>
    public static int GetMaxPlayerId(IReadOnlyList<Club> clubs)
    {
        return clubs.SelectMany(c => c.Team.Players).Max(p => p.Id);
    }

    private static PlayerAttributes GenerateYouthAttributes(
        IRng rng, Position pos, int baseAttr, int variance)
    {
        int Attr() => Math.Clamp(baseAttr + rng.NextInt(-variance, variance), 15, 70);
        int High() => Math.Clamp(baseAttr + 8 + rng.NextInt(-variance / 2, variance), 20, 75);
        int Low() => Math.Clamp(baseAttr - 5 + rng.NextInt(-variance, variance / 2), 10, 55);

        return pos switch
        {
            Position.GK => new PlayerAttributes
            {
                Finishing = Low(), Passing = Low(), Dribbling = Low(),
                FirstTouch = Attr(), Technique = Low(),
                Decisions = Attr(), Composure = High(), Positioning = Attr(),
                Anticipation = Attr(), OffTheBall = Low(),
                Speed = Attr(), Acceleration = Attr(), Stamina = Attr(),
                Strength = Attr(), Agility = Attr(),
                Consistency = Attr(), Leadership = Attr(), Flair = Low(), BigMatches = Attr(),
                Reflexes = High(), Handling = High(), GkPositioning = High(), Aerial = High()
            },
            Position.CB or Position.LB or Position.RB => new PlayerAttributes
            {
                Finishing = Low(), Passing = Attr(), Dribbling = Low(),
                FirstTouch = Attr(), Technique = Attr(),
                Decisions = High(), Composure = High(), Positioning = High(),
                Anticipation = High(), OffTheBall = Attr(),
                Speed = pos is Position.LB or Position.RB ? High() : Attr(),
                Acceleration = Attr(), Stamina = High(),
                Strength = High(), Agility = Attr(),
                Consistency = Attr(), Leadership = Attr(), Flair = Low(), BigMatches = Attr(),
                Reflexes = Low(), Handling = Low(), GkPositioning = Low(), Aerial = High()
            },
            Position.ST or Position.CF => new PlayerAttributes
            {
                Finishing = High(), Passing = Attr(), Dribbling = High(),
                FirstTouch = High(), Technique = High(),
                Decisions = Attr(), Composure = High(), Positioning = Attr(),
                Anticipation = Attr(), OffTheBall = High(),
                Speed = High(), Acceleration = High(), Stamina = Attr(),
                Strength = Attr(), Agility = High(),
                Consistency = Attr(), Leadership = Low(), Flair = High(), BigMatches = Attr(),
                Reflexes = Low(), Handling = Low(), GkPositioning = Low(), Aerial = Attr()
            },
            Position.LW or Position.RW => new PlayerAttributes
            {
                Finishing = Attr(), Passing = Attr(), Dribbling = High(),
                FirstTouch = High(), Technique = High(),
                Decisions = Attr(), Composure = Attr(), Positioning = Attr(),
                Anticipation = Attr(), OffTheBall = High(),
                Speed = High(), Acceleration = High(), Stamina = Attr(),
                Strength = Low(), Agility = High(),
                Consistency = Attr(), Leadership = Low(), Flair = High(), BigMatches = Attr(),
                Reflexes = Low(), Handling = Low(), GkPositioning = Low(), Aerial = Low()
            },
            _ => new PlayerAttributes
            {
                Finishing = Attr(), Passing = High(), Dribbling = Attr(),
                FirstTouch = High(), Technique = High(),
                Decisions = High(), Composure = High(), Positioning = Attr(),
                Anticipation = Attr(), OffTheBall = Attr(),
                Speed = Attr(), Acceleration = Attr(), Stamina = High(),
                Strength = Attr(), Agility = Attr(),
                Consistency = Attr(), Leadership = Attr(), Flair = Attr(), BigMatches = Attr(),
                Reflexes = Low(), Handling = Low(), GkPositioning = Low(), Aerial = Attr()
            }
        };
    }

    private static (IReadOnlyList<string> FirstNames, IReadOnlyList<string> LastNames) GetNamePool(
        string country)
    {
        return country switch
        {
            "Brasilândia" => (
                new[] { "Lucas", "Gabriel", "Matheus", "Rafael", "Pedro", "João", "Bruno", "Vinícius", "Kaio", "Enzo" },
                new[] { "Silva", "Santos", "Oliveira", "Souza", "Lima", "Costa", "Ferreira", "Almeida", "Pereira", "Nascimento" }),
            "Hispânia" => (
                new[] { "Carlos", "Diego", "Alejandro", "Pablo", "Miguel", "Sergio", "Álvaro", "Javier", "Hugo", "Adrián" },
                new[] { "García", "Rodríguez", "Martínez", "López", "Hernández", "Fernández", "Sánchez", "Pérez", "Ruiz", "Torres" }),
            "Angleterre" => (
                new[] { "James", "Oliver", "Harry", "Jack", "George", "Charlie", "Thomas", "William", "Daniel", "Samuel" },
                new[] { "Smith", "Jones", "Williams", "Brown", "Taylor", "Johnson", "Wilson", "Davies", "Robinson", "Thompson" }),
            "Itália Nova" => (
                new[] { "Marco", "Luca", "Alessandro", "Francesco", "Lorenzo", "Matteo", "Andrea", "Simone", "Giuseppe", "Davide" },
                new[] { "Rossi", "Russo", "Ferrari", "Esposito", "Bianchi", "Romano", "Colombo", "Ricci", "Marino", "Greco" }),
            _ => (
                new[] { "Alex", "Max", "Leo", "Tom", "Sam", "Ben", "Dan", "Chris", "Nick", "Ryan" },
                new[] { "Young", "Green", "White", "Black", "Grey", "Stone", "Wood", "Field", "Brook", "Hill" })
        };
    }
}

File: engine/transfers/youth-academy.ts
import { Position } from "@/engine/enums";
import type { Club, Player, PlayerAttributes } from "@/engine/models";
import type { IRng } from "@/engine/simulation/rng";

/**
 * Generates youth prospects for the academy recruitment system (card-based gacha).
 */

const YOUTH_POSITIONS: Position[] = [
  Position.GK,
  Position.CB,
  Position.LB,
  Position.RB,
  Position.CM,
  Position.CDM,
  Position.CAM,
  Position.LW,
  Position.RW,
  Position.ST,
  Position.CF,
];

/**
 * Generates 3 youth prospect cards. Player picks 1.
 */
export function generateProspects(
  rng: IRng,
  country: string,
  nextPlayerId: number,
): Array<{ prospect: Player; fee: number }> {
  const names = getNamePool(country);
  const prospects: Array<{ prospect: Player; fee: number }> = [];

  for (let i = 0; i < 3; i++) {
    const age = rng.nextInt(16, 19);
    const pos = YOUTH_POSITIONS[rng.nextInt(0, YOUTH_POSITIONS.length - 1)];

    const baseAttr = rng.nextInt(30, 55);
    const variance = 12;

    const firstName = names.firstNames[rng.nextInt(0, names.firstNames.length - 1)];
    const lastName = names.lastNames[rng.nextInt(0, names.lastNames.length - 1)];
    const name = `${firstName[0]}. ${lastName}`;

    const attrs = generateYouthAttributes(rng, pos, baseAttr, variance);

    const prospect: Player = {
      id: nextPlayerId + i,
      name,
      primaryPosition: pos,
      secondaryPosition: null,
      age,
      morale: rng.nextInt(50, 70),
      chemistry: rng.nextInt(30, 50),
      attributes: attrs,
      traits: [],
    };

    const ovr =
      pos === Position.GK
        ? goalkeeperOvrFromAttrs(attrs)
        : outfieldOvrFromAttrs(attrs);
    let fee = Math.round((ovr * 200) / 1000) * 1000;
    fee = Math.max(5_000, Math.min(15_000, fee));

    prospects.push({ prospect, fee });
  }

  return prospects;
}

/**
 * Returns the maximum player ID across all clubs (for generating new IDs).
 */
export function getMaxPlayerId(clubs: readonly Club[]): number {
  let maxId = 0;
  for (const club of clubs) {
    for (const player of club.team.players) {
      if (player.id > maxId) maxId = player.id;
    }
  }
  return maxId;
}

function generateYouthAttributes(
  rng: IRng,
  pos: Position,
  baseAttr: number,
  variance: number,
): PlayerAttributes {
  const attr = (): number =>
    clamp(baseAttr + rng.nextInt(-variance, variance), 15, 70);
  const high = (): number =>
    clamp(
      baseAttr + 8 + rng.nextInt(-Math.floor(variance / 2), variance),
      20,
      75,
    );
  const low = (): number =>
    clamp(
      baseAttr - 5 + rng.nextInt(-variance, Math.floor(variance / 2)),
      10,
      55,
    );

  switch (pos) {
    case Position.GK:
      return {
        finishing: low(),
        passing: low(),
        dribbling: low(),
        firstTouch: attr(),
        technique: low(),
        decisions: attr(),
        composure: high(),
        positioning: attr(),
        anticipation: attr(),
        offTheBall: low(),
        speed: attr(),
        acceleration: attr(),
        stamina: attr(),
        strength: attr(),
        agility: attr(),
        consistency: attr(),
        leadership: attr(),
        flair: low(),
        bigMatches: attr(),
        reflexes: high(),
        handling: high(),
        gkPositioning: high(),
        aerial: high(),
      };
    case Position.CB:
    case Position.LB:
    case Position.RB:
      return {
        finishing: low(),
        passing: attr(),
        dribbling: low(),
        firstTouch: attr(),
        technique: attr(),
        decisions: high(),
        composure: high(),
        positioning: high(),
        anticipation: high(),
        offTheBall: attr(),
        speed:
          pos === Position.LB || pos === Position.RB ? high() : attr(),
        acceleration: attr(),
        stamina: high(),
        strength: high(),
        agility: attr(),
        consistency: attr(),
        leadership: attr(),
        flair: low(),
        bigMatches: attr(),
        reflexes: low(),
        handling: low(),
        gkPositioning: low(),
        aerial: high(),
      };
    case Position.ST:
    case Position.CF:
      return {
        finishing: high(),
        passing: attr(),
        dribbling: high(),
        firstTouch: high(),
        technique: high(),
        decisions: attr(),
        composure: high(),
        positioning: attr(),
        anticipation: attr(),
        offTheBall: high(),
        speed: high(),
        acceleration: high(),
        stamina: attr(),
        strength: attr(),
        agility: high(),
        consistency: attr(),
        leadership: low(),
        flair: high(),
        bigMatches: attr(),
        reflexes: low(),
        handling: low(),
        gkPositioning: low(),
        aerial: attr(),
      };
    case Position.LW:
    case Position.RW:
      return {
        finishing: attr(),
        passing: attr(),
        dribbling: high(),
        firstTouch: high(),
        technique: high(),
        decisions: attr(),
        composure: attr(),
        positioning: attr(),
        anticipation: attr(),
        offTheBall: high(),
        speed: high(),
        acceleration: high(),
        stamina: attr(),
        strength: low(),
        agility: high(),
        consistency: attr(),
        leadership: low(),
        flair: high(),
        bigMatches: attr(),
        reflexes: low(),
        handling: low(),
        gkPositioning: low(),
        aerial: low(),
      };
    default:
      // Midfield positions (CM, CDM, CAM, LM, RM)
      return {
        finishing: attr(),
        passing: high(),
        dribbling: attr(),
        firstTouch: high(),
        technique: high(),
        decisions: high(),
        composure: high(),
        positioning: attr(),
        anticipation: attr(),
        offTheBall: attr(),
        speed: attr(),
        acceleration: attr(),
        stamina: high(),
        strength: attr(),
        agility: attr(),
        consistency: attr(),
        leadership: attr(),
        flair: attr(),
        bigMatches: attr(),
        reflexes: low(),
        handling: low(),
        gkPositioning: low(),
        aerial: attr(),
      };
  }
}

function getNamePool(country: string): {
  firstNames: readonly string[];
  lastNames: readonly string[];
} {
  switch (country) {
    case "Brasil":
      return {
        firstNames: [
          "Lucas",
          "Gabriel",
          "Matheus",
          "Rafael",
          "Pedro",
          "João",
          "Bruno",
          "Vinícius",
          "Kaio",
          "Enzo",
        ],
        lastNames: [
          "Silva",
          "Santos",
          "Oliveira",
          "Souza",
          "Lima",
          "Costa",
          "Ferreira",
          "Almeida",
          "Pereira",
          "Nascimento",
        ],
      };
    case "España":
      return {
        firstNames: [
          "Carlos",
          "Diego",
          "Alejandro",
          "Pablo",
          "Miguel",
          "Sergio",
          "Álvaro",
          "Javier",
          "Hugo",
          "Adrián",
        ],
        lastNames: [
          "García",
          "Rodríguez",
          "Martínez",
          "López",
          "Hernández",
          "Fernández",
          "Sánchez",
          "Pérez",
          "Ruiz",
          "Torres",
        ],
      };
    case "England":
      return {
        firstNames: [
          "James",
          "Oliver",
          "Harry",
          "Jack",
          "George",
          "Charlie",
          "Thomas",
          "William",
          "Daniel",
          "Samuel",
        ],
        lastNames: [
          "Smith",
          "Jones",
          "Williams",
          "Brown",
          "Taylor",
          "Johnson",
          "Wilson",
          "Davies",
          "Robinson",
          "Thompson",
        ],
      };
    case "Italia":
      return {
        firstNames: [
          "Marco",
          "Luca",
          "Alessandro",
          "Francesco",
          "Lorenzo",
          "Matteo",
          "Andrea",
          "Simone",
          "Giuseppe",
          "Davide",
        ],
        lastNames: [
          "Rossi",
          "Russo",
          "Ferrari",
          "Esposito",
          "Bianchi",
          "Romano",
          "Colombo",
          "Ricci",
          "Marino",
          "Greco",
        ],
      };
    default:
      return {
        firstNames: [
          "Alex",
          "Max",
          "Leo",
          "Tom",
          "Sam",
          "Ben",
          "Dan",
          "Chris",
          "Nick",
          "Ryan",
        ],
        lastNames: [
          "Young",
          "Green",
          "White",
          "Black",
          "Grey",
          "Stone",
          "Wood",
          "Field",
          "Brook",
          "Hill",
        ],
      };
  }
}

function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value));
}

/** Quick outfield OVR from raw attributes. */
function outfieldOvrFromAttrs(a: PlayerAttributes): number {
  return (
    (a.finishing +
      a.passing +
      a.dribbling +
      a.firstTouch +
      a.technique +
      a.decisions +
      a.composure +
      a.positioning +
      a.anticipation +
      a.offTheBall +
      a.speed +
      a.acceleration +
      a.stamina +
      a.strength +
      a.agility +
      a.consistency +
      a.leadership +
      a.flair +
      a.bigMatches) /
    19
  );
}

/** Quick GK OVR from raw attributes. */
function goalkeeperOvrFromAttrs(a: PlayerAttributes): number {
  return (
    (a.reflexes +
      a.handling +
      a.gkPositioning +
      a.aerial +
      a.composure +
      a.decisions +
      a.positioning +
      a.leadership +
      a.consistency) /
    9
  );
}
